src/simple_cache_client.rs

```diff
@@ -1873,6 +1873,137 @@ impl SimpleCacheClient {
+    ///             }
+    ///         }
+    ///         Elements::Values(_) => {
+    ///             panic!("should have returned elements with scores");
+    ///         }
+    ///     }
+    ///     None => {
+    ///         println!("sorted set or elements not found");
+    ///     }
+    /// }
+    /// # Ok(())
+    /// # })
+    /// # }
+    /// ```
+    pub async fn sorted_set_fetch_by_index(
+        &mut self,
+        cache_name: &str,
+        set_name: impl IntoBytes,
+        order: sorted_set::Order,
+        range: impl RangeBounds<i32>,
+        with_scores: bool,
+    ) -> MomentoResult<Option<crate::sorted_set::Elements>> {
+        // transforms the Rust `Range` start into a Momento start index. Because
+        // the Momento start index is always Inclusive (or Unbounded) we need to
+        // map the value of a Rust `Excluded` start bound to an
+        // `InclusiveStartIndex` by adding one to the value.
+        let start = match range.start_bound() {
+            std::ops::Bound::Included(v) => {
+                sorted_set_fetch_request::by_index::Start::InclusiveStartIndex(*v)
+            }
+            std::ops::Bound::Excluded(v) => {
+                sorted_set_fetch_request::by_index::Start::InclusiveStartIndex(*v + 1)
+            }
+            std::ops::Bound::Unbounded => {
+                sorted_set_fetch_request::by_index::Start::UnboundedStart(Unbounded {})
+            }
+        };
+
+        // transforms the Rust `Range` end into a Momento end index. Because the
+        // Momento end index is always Exclusive (or Unbounded) we need to map
+        // the value of a Rust `Included` end bound to an `ExclusiveEndIndex` by
+        // adding one to the value.
+        let end = match range.end_bound() {
+            std::ops::Bound::Included(v) => {
+                sorted_set_fetch_request::by_index::End::ExclusiveEndIndex(*v + 1)
+            }
+            std::ops::Bound::Excluded(v) => {
+                sorted_set_fetch_request::by_index::End::ExclusiveEndIndex(*v)
+            }
+            std::ops::Bound::Unbounded => {
+                sorted_set_fetch_request::by_index::End::UnboundedEnd(Unbounded {})
+            }
+        };
+
+        let request = self.prep_request(
+            cache_name,
+            SortedSetFetchRequest {
+                set_name: set_name.into_bytes(),
+                order: order.into(),
+                with_scores,
+                range: Some(sorted_set_fetch_request::Range::ByIndex(
+                    sorted_set_fetch_request::ByIndex {
+                        start: Some(start),
+                        end: Some(end),
+                    },
+                )),
+            },
+        )?;
+
+        let response = self
+            .data_client
+            .sorted_set_fetch(request)
+            .await?
+            .into_inner();
+
+        // this flattens the response returning a None for both a missing sorted
```

Comment:
I'm not sure if this is the behavior we want, but the ergonomics feel funny otherwise. Do we ever really return a Found<None> instead of a Found<Some(elements)> where elements contains a zero length vector of elements? If so, we do lose information here, but otherwise I think this flattening is beneficial.

Reply:
@cprice404 - it's actually a similar issue here. We have Found(Option<Vec<_>>) or Missing. It's unclear to me if there's some meaningful difference between the protocol crate response being a Missing or a Found(None) or a Found(empty vec). Again, we can hide this from the users, but I'm finding it hard to use the protocol crate to add this functionality to the client given the ambiguities.

I'm personally not finding the Node or Java SDKs to be super helpful as references.

Reply:
so, the response type from this should end up being either `Hit` or `Miss` (like Kenny's DictionaryFetch implementation).

`Miss` would mean that the collection does not exist (and I think corresponds to `Missing` in the protos if I am reading your comment correctly).

`Hit` means that the collection exists, but the vector could be empty.

I think that the reason that VluesWithScores vec could be optional in the proto is because we did `oneof _ValuesWithScores / _Values`, corresponding to the `with_scores` boolean on the request object.

I think that was done on the server side before we discussed the public / SDK API, and we have since decided that we do not want/need to support that `with_scores` option. Thus, in the implementation, with_scores can be hard-coded to true, and we would always expect the Option for ValuesWithScores to be populated. If it is not, that would be considered a bug/error.

Does that help?